### ISSUE TITLE
Colour output doesn't work in cmd on Windows

### ISSUE THREAD

**Issue by garyb** [purescript/spago#749]
It just prints the colour codes as text:

```
←[34m[info] ←[0mInstalling 65 dependencies.←[0m
←[34m[info] ←[0mSearching for packages cache metadata..←[0m
←[34m[info] ←[0mUnable to find packages cache metadata, downloading from GitHub..←[0m
←[34m[info] ←[0mInstalling "exists"←[0m
←[34m[info] ←[0mInstalling "distributive"←[0m
```

Perhaps related to #729, but the suggestion there is there's a regression in an upstream library, and it's been this way for as long as I've been using spago. That doesn't necessarily rule out the upstream regression, as I came to it late!

**Comment by garyb** [comment]
I probably should have mentioned, this isn't true for all commands. The `[info]` highlight works correctly for builds, for example.

**Comment by f-f** [comment]
>  The `[info]` highlight works correctly for builds, for example.

Oh, that's funny then - if all color was wrong then we'd know that there's something wrong with the lway we do logging, but if only _some_ coloring is wrong then I am not really sure how to even debug this. Do you have any ideas? We're using the [logging machinery from `RIO`](https://hackage.haskell.org/package/rio-0.1.20.0/docs/RIO.html#g:5)

**Comment by garyb** [comment]
Ah! I think it's something that `purs` is doing that makes it work actually. Any command that invokes the compiler has colour codes working correctly after "Compiling..." messages have been printed. `spago docs` is a perfect example actually, because it does this kind of thing:

```
←[34m[info] ←[0mGenerating documentation for the project. This might take a while...←[0m
Compiling documentation for Data.NaturalTransformation
Compiling documentation for Unsafe.Coerce
Compiling documentation for Record.Unsafe
Compiling documentation for Type.Proxy
... snip ...
Documentation written to: generated-docs/html
[info] Docs generation succeeded.
[info] Making the documentation searchable...
```

So before the compiler is invoked: no colour, after, colour.

**Comment by garyb** [comment]
I have an idea of something to try 🙂

**Comment by f-f** [comment]
Oh I see, that makes sense. Note that we just disable ANSI codes when `TERM=win`:

https://github.com/purescript/spago/blob/72315e003857e2a6c6e5904716d440925990ec21/src/Spago/RunEnv.hs#L24-L27

..but maybe we should just disable it when running on Windows at all?

**Comment by stkb** [comment]
Yeah would nice to get this fixed. I've tried to do some investigating but it's quite laborious with the little knowledge I have.

I've found though that if you just open start a stack repl in the Spago folder, and then do `System.IO.putStrLn "\ESC[31mRED\ESC[0m"`, then it works. But if you do that within the Spago app itself, it doesn't.

Also about `TERM=win`, I'm not sure what environments exactly that's for? Because that variable isn't set in vanilla Windows.

**Comment by garyb** [comment]
I've not yet managed to determine what the difference is between `purs` and `spago` that means it works in one place, and not the other, and that invoking `purs` inside `spago` will fix it from that point on too.

I was looking into how stdio was configured with utf-8 support, etc. to see if there was something there as it seems like it must be something along those lines, but I've not gotten back to it to try again yet. I'm probably going to try in a fresh CLI app entirely and see if it works "out of the box", and if not, I can maybe figure something out in there then take it back to `spago`.

**Comment by stkb** [comment]
> I'm probably going to try in a fresh CLI app entirely and see if it works "out of the box"

Yeah that was just what I was trying next, just with `stack new testproject` and it does indeed work fine there too.

**Comment by garyb** [comment]
Interesting! When I hacked spago to the point of printing a colour-coded line from its `main` before it does anything else (before `withUtf8` even), it still wasn't displaying correctly.

**Comment by stkb** [comment]
> with `stack new testproject` and it does indeed work fine there too.

Ok no I have to correct that. It's when running the exe through `stack exec` that the colors work. Works for a new project, works for Spago. When just running the .exe from the command line, it doesn't (also for a fresh project)

**Comment by f-f** [comment]
Could it be because the compiler [calls](https://github.com/purescript/purescript/blob/bc898395c2da556c8ec6b7c702a3dff6d2f1bb9c/app/Main.hs#L32) `IO.hSetBuffering IO.stdout IO.LineBuffering` but we don't?

**Comment by garyb** [comment]
Replicating the IO setup there was the first thing I tried when I hacked my local `spago` :smile:, I tried doing all that explicitly and avoiding `withUtf8`, and variations on all that without much luck. I really thought it would be the `LineBuffering` thing when I saw that.

**Comment by stkb** [comment]
That didn't work but I think it lead me to the right answer :) Just preparing a PR...

**Comment by stkb** [comment]
Oh I thought this would be closed when #768 was merged. I guess because I only mentioned the issue in the commit message body and not the first line. Anyway, closing now
